### ISSUE TITLE
aws-cdk-lib/aws-lambda-nodejs: cache esbuild results

### ISSUE THREAD

**Issue by bestickley** [aws/aws-cdk#26020]
### Describe the feature

The `NodejsFunction` construct should intelligently cache build results of esbuild and reuse them on subsequent deploys.

```ts
new NodejsLambda(this, "CachedNodejsLambda", { bundling: { cache: true } });
```

### Use Case

When working on a CDK app with many lambdas, deployments can take longer than I'd like. I want this to be faster so that the CDK provides a better DX and faster deployments. Work smarter, not harder, right? ;)

### Proposed Solution

1. Use [es-module-lexer](https://github.com/guybedford/es-module-lexer) to find all files imported by entry point file.
2. Compute hash of those files.
3. Use hash for `bundling.assetHash` of construct
4. Enjoy faster synths/deployments!

Technical considerations:
- Ensure source code from local workspaces (like PNPM) that change are included in hash

### Other Information

If this is out of scope of the AWS CDK (which I hope it is not), @NimmLorr has documented a solution using turbopack. See [this comment](https://github.com/aws/aws-cdk/issues/24456#issuecomment-1574158364).

### Acknowledgements

- [X] I may be able to implement this feature request
- [X] This feature might incur a breaking change

### CDK version used

N/A

### Environment details (OS name and version, etc.)

N/A

**Comment by pahud** [comment]
We had a similar discussion in 2020 https://github.com/aws/aws-cdk/issues/10286 and the conclusion was not to include additional npm modules and use docker instead. But welcome discussion if it's still relevant.

**Comment by tmokmss** [comment]
What if we just allow to set `assetHashType` as `SOURCE`? It would skip bundling when source files are not changed.

https://github.com/aws/aws-cdk/blob/0a61edf3499aa1a72709131d67ac849107870065/packages/aws-cdk-lib/aws-lambda-nodejs/lib/bundling.ts#L63

In PythonFunction, hash type is SOURCE by default, but idk why it isn't in NodejsFunction.

**Comment by bestickley** [comment]
@tmokmss, does `assetHashType` impact whether or not the function is bundled? I thought it was only for uploading the asset to S3? I'm asking for more of a "bundleHash"

**Comment by tmokmss** [comment]
@bestickley Yes if the asset hash is the same (if there is already a bundled result directory for the hash), bundling is skipped. 

https://github.com/aws/aws-cdk/blob/0a61edf3499aa1a72709131d67ac849107870065/packages/aws-cdk-lib/core/lib/asset-staging.ts#L429

**Comment by bestickley** [comment]
I'm still looking for time to dedicate to this, but wanted to document. I found this library which could do all heavy lifting of finding dependency tree: https://github.com/dependents/node-dependency-tree

**Comment by fab-mindflow** [comment]
I believe this should be prioritized with #24456 to address very slow builds in large CDK projects with lambdas.

**Comment by piotrmoszkowicz** [comment]
I would love to have that feature within CDK, it takes ages to build our Lambda dependant stacks!

**Comment by mhyland-phoenicia** [comment]
This solution doesn't nessesarilly deal with caching but, In the meantime a potential workaround that I ended up implementing was to create a prebuild step to bundle all the lambdas in parallel and then use Lambda.Code.fromAsset. We were able to shorten lambda bundling from 50+ seconds to<1 sec.
https://gist.github.com/mhyland-phoenicia/c16ed0907c264fc767215e6cb214e5ef

**Comment by whitakersp-fineos** [comment]
Would love to have this feature.  Our builds are super slow because of lambda building after adding powertools and prisma to our lambdas.

**Comment by LeoLapworthKT** [comment]
> Would love to have this feature. Our builds are super slow because of lambda building after adding powertools and prisma to our lambdas.

 Wondering if https://github.com/CloudSnorkel/cdk-turbo-layers helps?

**Comment by whitakersp-fineos** [comment]
> > Would love to have this feature. Our builds are super slow because of lambda building after adding powertools and prisma to our lambdas.
> 
> Wondering if https://github.com/CloudSnorkel/cdk-turbo-layers helps?

That's not using NodeJsFunction which provides a lot more capability than just function

**Comment by LeoLapworthKT** [comment]
> > Wondering if https://github.com/CloudSnorkel/cdk-turbo-layers helps?
> 
> That's not using NodeJsFunction which provides a lot more capability than just function

We use tubro-layers to bundle 3rd party dependencies, not the lambda function it self, into a layer (which is built in Cloudformation and only builds if changes in dependencies), we ALSO extract which packages are in that and supply to `NodeJsFunction`:

```ts
      layers: [ thePackagerLayer ],
      bundling: {
        externalModules,
      },
```

We are still using NodeJsFunction, but don't have the deploy overhead (if our dependencies haven't changed) of building 3rd party packages again.

So doesn't solve caching building of your node functions, but does remove the building of the dependencies.

We also added a `BUILD_STACK` env, which if set _only_ builds our core stacks + those which match this value, to minimise what CDK looks at, this has helped with deploying in development, though we added it before the `watch` and `hotdeploy` existed we still use it even with those

If anyone is interested I can see about putting the code we wrap Turbo-Layers with somewhere public

**Comment by ShivamJoker** [comment]
Bundling is too slow right now.
We need to get this merged.

**Comment by AllanOricil** [comment]
Is this going to be improved this year?

**Comment by AllanOricil** [comment]
I added `assetHash: cdk.AssetHashType.SOURCE` property in bundling and now my lambda functions are not rebuilt on every deploy. 

So, can't this issue be closed?

**Comment by AllanOricil** [comment]
After changing my lambda code, and rebuilding its cdk stack, no new asset was bundled. Is there an open issue for it?

**Comment by AllanOricil** [comment]
I made a mistake. `assetHashType` does not exist in NodeJsFunction

**Comment by AllanOricil** [comment]
I was able to cache my builds using `assetHash`

1. Compute hash using esbuild for all your lambda entrypoints and store it in a file. I stored it as a json object, where the key is the lambda entrypoint path and the value is the computed hash.

2. In your cdk project, set `assetHash` of your lambda to the hash found in the file you created in step 1. Use the entrypoint path to get the hash.

3. after the first bundling, store `cdk.out` somewhere you can retrieve in your CI automation. I stored it in a s3 bucket, in aws. On every new build, update your cached cdk.out if it is dirty (has diffs). Clear the cache folder every now and then, or when you want to have a full build.

**Comment by ShivamJoker** [comment]
Can you share any code sample for this? @AllanOricil

**Comment by AllanOricil** [comment]
@ShivamJoker this is the script I use to compute hashes to my lambdas.  My api is made of lambdas located at `./resources/lambda/handlers/api/controllers`.

````js
/* eslint-disable */
const fs = require("fs");
const crypto = require("crypto");
const esbuild = require("esbuild");
const path = require("path");
const pkg = JSON.parse(fs.readFileSync("package.json", "utf-8"));

// Computes the hash
async function computeHash(entryPoint) {
  const result = await esbuild.build({
    platform: "node",
    entryPoints: [entryPoint],
    write: false,
    bundle: true,
    treeShaking: true,
    minify: true,
    external: Object.keys(pkg.dependencies),
  });
  const hash = crypto.createHash("sha256");
  hash.update(Buffer.from(result.outputFiles[0].contents));
  return hash.digest("hex");
}

function walkDir(dir, fileList = []) {
  const files = fs.readdirSync(dir);

  files.forEach(function (file) {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);

    if (stat.isDirectory()) {
      fileList = walkDir(filePath, fileList);
    } else if (path.basename(file) === "index.ts") {
      fileList.push(filePath);
    }
  });

  return fileList;
}

(async () => {
  const dir = "./resources/lambda/handlers/api/controllers";
  const entryPoints = walkDir(dir);
  let hashes = {};

  for (let ep of entryPoints) {
    hashes[ep] = await computeHash(ep);
  }

  fs.writeFileSync(
    "./resources/lambda/computed-hashes.json",
    JSON.stringify(hashes, null, 2),
  );
})();
````


This is how I use the hashes located at `./resources/lambda/computed-hashes.json` in the CDK when creating my lambdas:

````js
private createApiMethodLambdaFunction(
    name: string,
    entryPath: string,
    environment?: environment,
    permissions?: string[],
    secrets?: cdk.aws_secretsmanager.ISecret[],
    database?: cdk.aws_rds.IDatabaseInstance,
  ): NodejsFunction {
    const entryPoint = path.join(
      DEFAULT_HANDLERS_DIRECTORY_PATH,
      entryPath,
      "/index.ts",
    );
    const relativePath = path.relative(process.cwd(), entryPoint);
    const assetHash = this.computedLambdaHashes?.[relativePath];

    const lambdaFunction = new NodejsFunction(this, `${name}-api-function`, {
      functionName: `${name}-api-function`,
      entry: entryPoint,
      handler: "main",
      runtime: lambda.Runtime.NODEJS_18_X,
      timeout: cdk.Duration.seconds(60),
      memorySize: cdk.Size.mebibytes(128).toMebibytes(),
      depsLockFilePath: path.resolve(__dirname, "../../package-lock.json"),
      retryAttempts: 0,
      logRetention: cdk.aws_logs.RetentionDays.ONE_WEEK,
      environment: {
        NODE_ENV: process.env.NODE_ENV ?? "production",
        LOG_LEVEL: process.env.LOG_LEVEL ?? "info",
        ...environment,
      },
      bundling: {
        sourceMap: true,
        assetHash: assetHash ?? undefined,
        nodeModules: [
          "winston",
          "@prisma/client",
          "@sentry/serverless",
          "ajv",
          "crypto-js",
          "axios",
          "axios-retry",
          "uuid",
        ],
        externalModules: ["aws-sdk", "@aws-sdk"],
        user: "root",
        forceDockerBundling: true,
        dockerImage: cdk.DockerImage.fromRegistry(
          getEnvVar("LAMBDA_DOCKER_IMAGE"),
        ),
        commandHooks: {
          beforeBundling(i: string, o: string): string[] {
            return ['echo "BEFORE BUNDLING"', `echo "${name}-api-function"`];
          },
          beforeInstall(i: string, o: string): string[] {
            return ['echo "BEFORE INSTALL"', `cp -r ${i}/prisma ${o}`];
          },
          afterBundling(i: string, o: string): string[] {
            return [
              'echo "AFTER BUNDLING"',
              `cp -r ${i}/node_modules/.prisma node_modules`,
              `cp ${i}/.sentryclirc .`,
              `sentry-cli sourcemaps inject --org journeylab --project node-awslambda .`,
              `sentry-cli sourcemaps upload --org journeylab --project node-awslambda .`,
              `rm .sentryclirc`,
            ];
          },
        },
      },
      vpc: this.vpc,
    });

    if (permissions !== undefined && permissions.length >= 0) {
      lambdaFunction.addToRolePolicy(
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: permissions,
          resources: ["*"],
        }),
      );
    }

    if (secrets !== undefined && secrets.length >= 0) {
      for (const secret of secrets) {
        secret.grantRead(lambdaFunction);
      }
    }

    if (database) {
      database.connections.allowDefaultPortFrom(lambdaFunction);
    }

    return lambdaFunction;
  }
````

**Comment by azhong-git** [comment]
+1 This plus https://github.com/aws/aws-cdk/issues/24456 would be super helpful for projects with many nodejs lambda functions.
